lago/subnet_lease.py

```diff
@@ -17,222 +17,505 @@
     """
-    Checs if the given lease still has a prefix that owns it
-
-    Args:
-        path (str): Path to the lease
-
-    Returns:
-        bool: ``True`` if the uuid path in the lease still exists and is the
-            same as the one in the lease
-    """
-    with open(path) as f:
-        uuid_path, uuid = json.load(f)
-
-    if not os.path.isfile(uuid_path):
-        return False
-
-    with open(uuid_path) as f:
-        return f.read() == uuid
 
-
-@_validate_lease_dir_present
-@_locked
-def _acquire(uuid_path):
-    """
-    Lease a free network for the given uuid path
-
-    Args:
-        uuid_path (str): Path to the uuid file of a :class:`lago.Prefix`
-
-    Returns:
-        int or None: the third element of the dotted ip of the leased network
-            or ``None`` if no lease was available
-
-    .. todo::
-        Raise exception or something instead of returning None so the
-        caller can handle the failure case
-    """
-
-    lease_dir = config['lease_dir']
-    for index in range(MIN_SUBNET, MAX_SUBNET + 1):
-        lease_file = os.path.join(lease_dir, '%d.lease' % index)
-        if os.path.exists(lease_file):
-            if _lease_valid(lease_file):
+    def __init__(
+        self,
+        path=None,
+        min_third_octet=200,
+        max_third_octet=255,
+    ):
+
+        self._path = path or config['lease_dir']
+        self._cidr = 24
+        self._subnet_template = '{}/{}'.format('192.168.{}.0', self._cidr)
+        self._min_third_octet = min_third_octet
+        self._max_third_octet = max_third_octet
+        self._min_subnet = IPNetwork(
+            self._subnet_template.format(min_third_octet)
+        )
+        self._max_subnet = IPNetwork(
+            self._subnet_template.format(max_third_octet)
+        )
+        self._validate_lease_dir()
+
+    def _create_lock(self):
+        return utils.LockFile(
+            path=os.path.join(self.path, LOCK_NAME), timeout=5
+        )
+
+    def _validate_lease_dir(self):
+        """
+        Validate that the directory used by this store exist,
+            otherwise create it.
+        """
+        if not os.path.isdir:
+            os.makedirs(self._path)
+
+    def acquire(self, uuid_path, subnet=None):
+        """
+        Lease a free subnet for the given uuid path.
+        If subnet is given, try to lease that subnet, otherwise try to lease a
+        free subnet.
+
+        Args:
+            uuid_path (str): Path to the uuid file of a :class:`lago.Prefix`
+            subnet (str): A subnet to lease.
+        Returns:
+            netaddr.IPAddress: An object which represents the subnet.
+
+        Raises:
+            LagoSubnetLeaseException:
+                1. If this store is full
+                2. If the requested subnet is already taken.
+            LagoSubnetLeaseLockException:
+                If the lock to self.path can't be acquired.
+        """
+        try:
+            with self._create_lock():
```

Comment:
The interface of the store doesn't provide a method for only reading its content, and even if it was, a lock should be created while reading, otherwise, the store can be changed while reading.